pennylane_lightning_gpu/src/util/MPIManager.hpp

```diff
@@ -0,0 +1,612 @@
+        size_ = static_cast<size_t>(size_int);
+
+        findVendor();
+        findVersion();
+        getNumProcsPerNode();
+        check_mpi_config();
+    }
+
+    MPIManager(int argc, char **argv) {
+        PL_MPI_IS_SUCCESS(MPI_Init(&argc, &argv));
+        isExternalComm_ = false;
+        communicator_ = MPI_COMM_WORLD;
+        int rank_int;
+        int size_int;
+        PL_MPI_IS_SUCCESS(MPI_Comm_rank(communicator_, &rank_int));
+        PL_MPI_IS_SUCCESS(MPI_Comm_size(communicator_, &size_int));
+
+        rank_ = static_cast<size_t>(rank_int);
+        size_ = static_cast<size_t>(size_int);
+
+        findVendor();
+        findVersion();
+        getNumProcsPerNode();
+        check_mpi_config();
+    }
+
+    MPIManager(const MPIManager &other) {
+        isExternalComm_ = true;
+        rank_ = other.rank_;
+        size_ = other.size_;
+        communicator_ = other.communicator_;
+        vendor_ = other.vendor_;
+        version_ = other.version_;
+        subversion_ = other.subversion_;
+        size_per_node_ = other.size_per_node_;
+    }
+
+    virtual ~MPIManager() final {
+        if (!isExternalComm_) {
+            int initflag;
+            int finflag;
+            PL_MPI_IS_SUCCESS(MPI_Initialized(&initflag));
+            PL_MPI_IS_SUCCESS(MPI_Finalized(&finflag));
+            if (initflag && !finflag) {
+                PL_MPI_IS_SUCCESS(MPI_Finalize());
+            }
+        }
+    }
+
+    // General MPI operations
+    /**
+     * @brief Get the process rank in the communicator.
+     */
+    auto getRank() const -> size_t { return rank_; }
+
+    /**
+     * @brief Get the process number in the communicator.
+     */
+    auto getSize() const -> size_t { return size_; }
+
+    /**
+     * @brief Get the number of processes per node in the communicator.
+     */
+    auto getSizeNode() const -> size_t { return size_per_node_; }
+
+    /**
+     * @brief Get the communicator.
+     */
+    MPI_Comm getComm() { return communicator_; }
+
+    /**
+     * @brief Get an elapsed time.
+     */
+    auto getTime() const { return MPI_Wtime(); }
+
+    /**
+     * @brief Get the MPI vendor.
+     */
+    auto getVendor() const -> std::string { return vendor_; }
+
+    /**
+     * @brief Get the MPI version.
+     */
+    auto getVersion() const -> std::tuple<size_t, size_t> {
+        return {version_, subversion_};
+    }
+
+    /**
+     * @brief Find the MPI vendor.
+     */
+    void findVendor() {
+        char version[MPI_MAX_LIBRARY_VERSION_STRING];
+        int resultlen;
+
+        PL_MPI_IS_SUCCESS(MPI_Get_library_version(version, &resultlen));
+
+        std::string version_str = version;
+
+        if (version_str.find("Open MPI") != std::string::npos) {
+            vendor_ = "Open MPI";
+        } else if (version_str.find("MPICH") != std::string::npos) {
+            vendor_ = "MPICH";
+        } else {
+            std::cerr << "Unsupported MPI implementation.\n";
+            exit(EXIT_FAILURE);
```

Comment:
Rather than calling `exit` it is best to raise an exception that we define to indicate the support is not there. Exit will kill the Python process, and anything else running, whereas with an exception we can catch it and aim to use a different device instead, and build some fallback logic.

Reply:
Sure!